lls/main.c

```diff
@@ -16,11 +16,203 @@
+
+	/* Wait for network devices to start. */
+	while (lls_conf->net->configuring)
+		;
+
+	/* TODO Add ARP and ND filters. */
+
+	while (likely(!exiting)) {
+		unsigned int num_changes = 0;
+
+		/* Read in packets on front and back interfaces. */
+		process_pkts(lls_conf, &lls_conf->net->front,
+			lls_conf->rx_queue_front, lls_conf->tx_queue_front);
+		if (lls_conf->net->back_iface_enabled)
+			process_pkts(lls_conf, &lls_conf->net->back,
+				lls_conf->rx_queue_back,
+				lls_conf->tx_queue_back);
+
+		/* Process any updates. */
+		num_changes = lls_cache_writer(lls_conf);
+
+		if (likely(num_changes == 0)) {
+			/*
+			 * If there are no requests to go through, then do a
+			 * scan of the cache (if enough time has passed).
+			 *
+			 * XXX In theory, many new LLS changes could starve
+			 * the ability to scan, but this will not likely
+			 * happen. In fact, we may want to reduce the amount
+			 * of times this is called, since reading the HPET
+			 * timer is inefficient. See the timer application.
+			 */
+			rte_timer_manage();
+		}
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"lls: the LLS block at lcore = %u is exiting\n",
+		lls_conf->lcore_id);
+
+	return cleanup_lls();
+}
+
+static int
+assign_lls_queue_ids(struct lls_config *lls_conf)
+{
+	int ret = get_queue_id(&lls_conf->net->front, QUEUE_TYPE_RX,
+		lls_conf->lcore_id);
+	if (ret < 0)
+		return ret;
+	lls_conf->rx_queue_front = (uint16_t)ret;
+
+	ret = get_queue_id(&lls_conf->net->front, QUEUE_TYPE_TX,
+		lls_conf->lcore_id);
+	if (ret < 0)
+		return ret;
+	lls_conf->tx_queue_front = (uint16_t)ret;
+
+	if (lls_conf->net->back_iface_enabled) {
+		ret = get_queue_id(&lls_conf->net->back, QUEUE_TYPE_RX,
+			lls_conf->lcore_id);
+		if (ret < 0)
+			return ret;
+		lls_conf->rx_queue_back = (uint16_t)ret;
+
+		ret = get_queue_id(&lls_conf->net->back, QUEUE_TYPE_TX,
+			lls_conf->lcore_id);
+		if (ret < 0)
+			return ret;
+		lls_conf->tx_queue_back = (uint16_t)ret;
+	}
+
+	return 0;
+}
+
+static void
+lls_scan(__attribute__((unused)) struct rte_timer *timer,
+	__attribute__((unused)) void *arg)
+{
+	/* TODO Scan the ARP and ND caches. */
+}
 
 int
-run_lls(void)
+run_lls(struct net_config *net_conf, struct lls_config *lls_conf)
 {
-	/* TODO Initialize and run Link Layer Support functional block. */
+	int ret;
+
+	if (net_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	lls_conf->net = net_conf;
+
+	/* Do LLS cache scan every LLS_CACHE_SCAN_INTERVAL seconds. */
+	rte_timer_init(&lls_conf->timer);
+	ret = rte_timer_reset(&lls_conf->timer,
+		LLS_CACHE_SCAN_INTERVAL * rte_get_timer_hz(), PERIODICAL,
+		lls_conf->lcore_id, lls_scan, lls_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, TIMER, "Cannot set LLS scan timer\n");
+		return -1;
+	}
+
+	ret = assign_lls_queue_ids(lls_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: cannot assign queues\n");
+		goto timer;
+	}
+
+	ret = init_mailbox("lls", MAILBOX_MAX_ENTRIES,
+		sizeof(struct lls_cache_req) + LLS_CACHE_MAX_KEY_LEN,
```

Comment:
You forgot to drop `LLS_CACHE_MAX_KEY_LEN` here after you changed field `ip_key` of `struct lls_cache_req`.